### ISSUE TITLE
Allow for server-side operations on authentication callback 

### ISSUE THREAD

**Issue by matthieualouis** [auth0/nextjs-auth0#83]
We'd trying out this library to add authentication to a new Next.js project. Everything seems to work nicely, however I think the callback handler can be improved.

When a user signs-up, and is redirected to `/api/callback` for the `callbackHandler` to perform the code exchange with Auth0, I would like to perform some operations on our database before the user is redirected back to the application homepage. 

This doesn't seem possible right now. In an ideal world, I'd be able to do something along these lines: 

```
import auth0 from '../../lib/auth0';
import User from '../entities/User';

export default async function callback(req, res) {
  try {
    const { user } = await auth0.exchangeCode(req)
    await User.createFromAuth0(user);
    auth0.redirect(res, { redirectTo: '/' });
  } catch (error) {
    console.error(error)
    res.status(error.status || 500).end(error.message)
  }
}
```

Is that something which you are planning to add, or is it something I can contribute to? 

Thanks, 

Matt

**Comment by HurricaneInteractive** [comment]
@matthieualouis the redirecting is baked into the `handleCallback` method ([See the source](https://github.com/auth0/nextjs-auth0/blob/master/src/handlers/callback.ts#L50)).

What you could try (might not be the nicest thing though), is set the `redirectTo` to a different `api` route and then fetch the [current user](https://github.com/auth0/nextjs-auth0#user-profile) and perform your database operations.

Another option is looking into [Auth0 hooks](https://auth0.com/blog/introducing-auth0-hooks/).

**Comment by remiroyc** [comment]
I'm agree with @matthieualouis request. I think that would be good if handleCallback can return ```session.user``` directly or take a callback function.

Because, as the auth0 doc specifies, I had to synchronize - after each logins - user data in my own database.

**Comment by kwiss** [comment]
@HurricaneInteractive Auth0 hooks, seems to be a sad solution since with that we can't deploy feature branches, staging and production environnement with differents databases

**Comment by derrickpelletier** [comment]
I did what @HurricaneInteractive suggested with overriding the `redirectTo` with _another_ callback page that handles the sync with our backend, then performs the intended original redirect. It works.

I've had to do hacky solutions to get around the redirects in `handleLogout` and `handleProfile` as well 😟

I do think there'd be value in being able to call _most_ of the methods here without implicit redirects. The only way to use this effectively creates a relatively hacky set of local endpoints...

**Comment by sandrinodimattia** [comment]
Added as part of [v0.12.0](https://github.com/auth0/nextjs-auth0/releases/tag/v0.12.0)

You can now inject logic right before the session is created. It can be used to control what goes in the session, call your own backend (eg: to provision a user), prevent the user from signing in, ...

```js
export default async function callback(req, res) {
  try {
    await auth0.handleCallback(req, res, {
      onUserLoaded: async (req, res, session, state) => {
        await User.createFromAuth0(user);
        return session;
      }
    });
  } catch (error) {
    console.error(error);
    res.status(error.status || 400).end(error.message);
  }
}
```

**Comment by derrickpelletier** [comment]
@sandrinodimattia is there a plan to add similar apis to the other handlers? Seems like a some of the other ones could benefit from this flexibility as well.

**Comment by sandrinodimattia** [comment]
Feel free to open separate issues if you have additional use cases.

**Comment by rebeccapeltz** [comment]
I've added this code to the callback event
```
 onUserLoaded: async (req, res, session, state) => {
        console.log("user loaded")
        return {
          ...session,
          user: {
            ...session.user
          }
        };
      },
      redirectTo: '/' })
```
as described.  The console log never appears. I've inserted `debugger` before and after the `handleCallback`  and the code stops so I know it's calling the handleCallback.  It also logs that it has called the handleCallback.  When I put the `debugger` after the onUserLoaded it never stops.  I'm concluding that it never fires this event.  Is there anything I need to do to make it fire the event?  This seems really useful and I want to get it working.  Logging in and out is working fine otherwiser.
